#Peer-graded Assignment: Making Data Management Decisions

Submit by February 11, 11:59 PM PST

## Important Information

It is especially important to submit this assignment before the deadline, February 11, 11:59 PM PST, because it must be graded by others. If you submit late, there may not be enough classmates around to review your work. This makes it difficult - and in some cases, impossible - to produce a grade. Submit on time to avoid these risks.

## Instructions

This week, you will be making and implementing decisions about data management for the variables that you have chosen to examine based on the steps reviewed in the videos and supplemental materials. This assignment is important because it offers you the opportunity to practice making sound data management decisions and think about how these decisions will impact your research.
Review criteria

Your assessment will be based on the evidence you provide that you have completed all of the steps. When relevant, gradients in the scoring will be available to reward clarity (for example, you will get one point for submitting output that is not understandable, but two points if it is understandable). In all cases, consider that the peer assessing your work is likely not an expert in the field you are analyzing. You will be assessed equally on your description of your frequency distributions.

Specific rubric items, and their point values, are as follows:

* Was the program output interpretable (i.e. organized and labeled)? (1 point)
* Does the program output display three data managed variables as frequency tables? (1 point)
* Did the summary describe the frequency distributions in terms of the values the variables take, how often they take them, the presence of missing data, etc.? (2 points)

## Instructions

Continue with the program you’ve successfully run.

Decide how you will manage your variables.

**STEP 1:** Make and implement data management decisions for the variables you selected.

Data management includes such things as coding out missing data, coding in valid data, recoding variables, creating secondary variables and binning or grouping variables. Not everyone does all of these, but some is required.

**STEP 2:** Run frequency distributions for your chosen variables and select columns, and possibly rows.

Your output should be interpretable (i.e. organized and labeled).

**WHAT TO SUBMIT:**

Once you have written a successful program that manages your data, create a blog entry where you post your program and the results/output that displays at least 3 of your data managed variables as frequency distributions. Write a few sentences describing these frequency distributions in terms of the values the variables take, how often they take them, the presence of missing data, etc.

To try:

* clean data with pandas.DataFrame.replace()
* use pandas.DataFrame.map([Dict]) to map old values to new
* provide secondary variables
* collapse multiple columns into one (done)
* use qcut (quartile cut) and cut to create categories
* use crosstab to check data
* other functions: isnull()

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Secondary Data Source:

vic = pd.read_csv('2016_census_vic_income.csv')
vic.head()

postcode gender  age_low  age_high  income_low  income_high  number
0  POA3000      M       15        19         1.0        149.0      78
1  POA3002      M       15        19         1.0        149.0       6
2  POA3003      M       15        19         1.0        149.0      17
3  POA3004      M       15        19         1.0        149.0      10
4  POA3005      M       15        19         1.0        149.0       0

## Revise the population distribution based on post code

The data here are already categorical. We divide the population by their
income brackets (income_low and income_high).

But I will explore if we can further explore data in different groupings.


In [3]:
# replace unknown data to nan
vic['income_low'] = vic['income_low'].replace(-1.0, np.nan)
vic['income_low'].value_counts()

 500.0     18846
 1750.0    18846
 1000.0    18846
 300.0     18846
 2000.0    18846
 400.0     18846
-99.0      18846
 1500.0    18846
 3000.0    18846
 650.0     18846
 1250.0    18846
 150.0     18846
 800.0     18846
 1.0       18846
Name: income_low, dtype: int64

In [4]:
# replace unknown data to nan
vic['income_low'] = vic['income_low'].replace(-99.0, np.nan)
vic['income_low'].value_counts()

500.0     18846
1750.0    18846
1000.0    18846
300.0     18846
2000.0    18846
400.0     18846
1500.0    18846
3000.0    18846
650.0     18846
1250.0    18846
150.0     18846
800.0     18846
1.0       18846
Name: income_low, dtype: int64

In [5]:
# replace unknown data to nan
vic['income_high'] = vic['income_high'].replace(-99.0, np.nan)
vic['income_high'].value_counts()

1999.0    18846
999.0     18846
499.0     18846
1749.0    18846
799.0     18846
399.0     18846
2999.0    18846
1499.0    18846
649.0     18846
9999.0    18846
1249.0    18846
299.0     18846
149.0     18846
0.0       18846
Name: income_high, dtype: int64

In [6]:
# replace unknown data to nan
vic['income_high'] = vic['income_high'].replace(9999.0, np.nan)
vic['income_high'].value_counts()

1999.0    18846
999.0     18846
499.0     18846
1749.0    18846
799.0     18846
399.0     18846
2999.0    18846
1499.0    18846
649.0     18846
1249.0    18846
299.0     18846
149.0     18846
0.0       18846
Name: income_high, dtype: int64

### Categorical data: Income bracket population in each post code
Here I discover that it is not always possible to cut the population into bins you want.
As you can see that there are way too many in the first group, 46%.

Australia (Victoria state)'s **population in each income bracket is very small when divided up by post code**.

In [7]:
income_pop_quartiles = pd.qcut(vic['number'], 10, duplicates='drop')
income_pop_quartiles.value_counts(sort=False)

(-0.001, 3.0]     130330
(3.0, 5.0]         16494
(5.0, 10.0]        25930
(10.0, 20.0]       26566
(20.0, 42.0]       26912
(42.0, 97.0]       28417
(97.0, 5254.0]     28041
Name: number, dtype: int64

In [8]:
income_pop_quartiles.value_counts(normalize=True)

(-0.001, 3.0]     0.461035
(42.0, 97.0]      0.100524
(97.0, 5254.0]    0.099193
(20.0, 42.0]      0.095200
(10.0, 20.0]      0.093976
(5.0, 10.0]       0.091726
(3.0, 5.0]        0.058347
Name: number, dtype: float64

In [9]:
# Females only

vic_f = vic[vic['gender'] == 'F']
f_income_pop_quartiles = pd.qcut(vic_f['number'], 10, duplicates='drop')
f_income_pop_quartiles.value_counts(sort=False, normalize=True)

(-0.001, 3.0]     0.503831
(3.0, 7.0]        0.096328
(7.0, 16.0]       0.106994
(16.0, 35.0]      0.094789
(35.0, 79.0]      0.098334
(79.0, 2985.0]    0.099724
Name: number, dtype: float64

In [10]:
# Males only

vic_m = vic[vic['gender'] == 'M']
m_income_pop_quartiles = pd.qcut(vic_m['number'], 10, duplicates='drop')
m_income_pop_quartiles.value_counts(sort=False, normalize=True)

(-0.001, 3.0]     0.491054
(3.0, 4.0]        0.037398
(4.0, 8.0]        0.083763
(8.0, 15.0]       0.090534
(15.0, 31.0]      0.097612
(31.0, 70.0]      0.100775
(70.0, 2265.0]    0.098864
Name: number, dtype: float64

## Secondary variables

* vic['income']: the average of the lower end and the higher end in each bracket
* vic['total_income']: vic['income'] * number of people. This field exists in the raw dataset but here we reconstruct it

In [11]:
vic['income'] = (vic['income_low'] + vic['income_high'])/2

In [12]:
vic['income'].value_counts()

1874.5    18846
899.5     18846
449.5     18846
224.5     18846
1624.5    18846
724.5     18846
349.5     18846
1374.5    18846
2499.5    18846
75.0      18846
574.5     18846
1124.5    18846
Name: income, dtype: int64

In [13]:
vic['total_income'] = vic['income'] * vic['number']
vic.head()

postcode gender  age_low  age_high  income_low  income_high  number  income  \
0  POA3000      M       15        19         1.0        149.0      78    75.0   
1  POA3002      M       15        19         1.0        149.0       6    75.0   
2  POA3003      M       15        19         1.0        149.0      17    75.0   
3  POA3004      M       15        19         1.0        149.0      10    75.0   
4  POA3005      M       15        19         1.0        149.0       0    75.0   

   total_income  
0        5850.0  
1         450.0  
2        1275.0  
3         750.0  
4           0.0

## Income total and per head via secondary variables

BY inspecting secondary varialbes, we can see the overall incomes. 

Male's incomes are higher than female's.

In [45]:
# Female population and income total and per head
vic_f = vic[vic['gender'] == 'F']
vic_f_pop = vic_f['number'].sum()
vic_f_total_income = vic_f['total_income'].sum()
print(f'Vic Female:\n\tPopulation: \t{vic_f_pop:>20,d}\n', 
      f'\tincome total: {vic_f_total_income:>20,.0f}\n',
      f'\taverage weekly income: {vic_f_total_income / vic_f_pop:>20.2f}')

Vic Female:
	Population: 	           2,480,205
 	income total:        1,467,123,144
 	average weekly income:               591.53


In [46]:
# Male population and income total and per head
vic_m = vic[vic['gender'] == 'M']
vic_m_pop = vic_m['number'].sum()
vic_m_total_income = vic_m['total_income'].sum()
print(f'Vic Male:\n\tPopulation: \t{vic_m_pop:>20,d}\n', 
      f'\tincome total: {vic_m_total_income:>20,.0f}\n',
      f'\taverage weekly income: {vic_m_total_income / vic_m_pop:>20.2f}')

Vic Male:
	Population: 	           2,339,649
 	income total:        1,810,145,426
 	average weekly income:               773.68


## Find out rich postal areas by secondary variables and groupby

In [16]:
vic_postal_income = vic[vic['gender'] == 'P'].groupby('postcode')['number', 'total_income'].sum()

In [17]:
vic_postal_income['avg_income'] = vic_postal_income['total_income'] / vic_postal_income['number']

In [18]:
vic_postal_income.nlargest(10, 'avg_income')

number  total_income   avg_income
postcode                                   
POA3785       28       29764.0  1063.000000
POA3662      771      747619.0   969.674449
POA3852       72       69268.5   962.062500
POA3067     7525     6738170.5   895.437940
POA3184    13411    12000147.0   894.798822
POA3121    27444    24471396.5   891.684758
POA3002     4631     4082104.5   881.473656
POA3015    13837    12134495.5   876.959999
POA3704       17       14891.5   875.970588
POA3013    12168    10538549.5   866.087237

In [19]:
vic_postal_income.nsmallest(10, 'avg_income')

number  total_income  avg_income
postcode                                  
POA3086      881      145962.5  165.678207
POA3800     2180      398334.5  182.722248
POA3050       76       14292.0  188.052632
POA3010     1581      316896.5  200.440544
POA3520      135       37603.0  278.540741
POA3506       15        4422.0  294.800000
POA3045       86       25743.0  299.337209
POA3026       40       12412.5  310.312500
POA3889      124       39204.0  316.161290
POA3482      111       38861.0  350.099099

## Mapping postcodes to suburb names

Australia Post Code Data: (not free) https://postcode.auspost.com.au/product_display.html?id=1

After digging around, the Australian Bureau of Statistics (ABS) says:

POSTAL AREA NAMES

Postal Areas (POAs) are not allocated names since there is no standardised name available for Australia Post postcodes. In most cases the code is repeated in the name field except where State or Territory (S/T) boundaries are crossed when a description is added.

It is contradicting to our common belief but reading through the process of building postcode into the dataset, I can understand POAs are not accurate enough for ABS.

Still, I think it's fairly useful information so I try to collect POAs from a different dataset:

[Victorian Electors by Locality, Postcode and Electorates](https://www.data.vic.gov.au/data/dataset/victorian-electors-by-locality-postcode-and-electorates)

But I am relectant to mapping the whole dataframe to this imperfect information.


In [20]:
vic_ele = pd.read_excel('../dataset_Vic/LocalityFinder.xls', skiprows=2)

In [21]:
vic_ele.columns

Index(['Locality Name', 'Post\r\nCode', 'Municipality\r\nName', 'Ward\r\nName',
       'LGA', 'District\r\nName', 'District\r\nCode', 'Region\r\nName',
       'Region\r\nCode', 'Property\r\nCount', 'Elector\r\nCount'],
      dtype='object')

In [22]:
vic_postcode_names = vic_ele[['Locality Name', 'Post\r\nCode']]
vic_postcode_names.columns = ['locality name', 'postcode']

Does the locality name and post code have 1-to-1 relationship?


In [23]:
len(vic_postcode_names['locality name'])

3881

In [24]:
print(f'Total postal records: {len(vic_postcode_names)}')
print(f'Unique post name records: {len(pd.unique(vic_postcode_names["locality name"]))}') 
print(f'Unique post code records" {len(pd.unique(vic_postcode_names["postcode"]))}')

Total postal records: 3881
Unique post name records: 2877
Unique post code records" 695


In [25]:
vic_postcode_names.groupby('postcode').count().head()

locality name
postcode               
3000                  1
3002                  1
3003                  1
3004                  4
3006                  3

In [26]:
vic_postcode_names[vic_postcode_names['postcode'] == 3004]

locality name  postcode
2192     Melbourne      3004
2193     Melbourne      3004
2194     Melbourne      3004
2195     Melbourne      3004

In [27]:
# remove duplicates
vic_postcode_names = vic_postcode_names.drop_duplicates()
len(pd.unique(vic_postcode_names['locality name'])), len(pd.unique(vic_postcode_names['postcode']))

(2877, 695)

In [28]:
vic_postcode_names['postcode'].dtypes

dtype('int64')

In [29]:
# This operation has SettingWithCopyWarning
# vic_postcode_names['postcode'] = 'POA' + vic_postcode_names['postcode'].astype(str)
vic_postcode_names['postcode'] = vic_postcode_names['postcode'].apply(lambda row: 'POA' + str(row))

In [30]:
vic_postcode_names.head()

locality name postcode
0      Abbeyard  POA3737
1    Abbotsford  POA3067
2    Aberfeldie  POA3040
3     Aberfeldy  POA3825
4       Acheron  POA3714

In [31]:
vic_postcode_names.columns

Index(['locality name', 'postcode'], dtype='object')

In [32]:
vic_postal_income.count()

number          698
total_income    698
avg_income      697
dtype: int64

In [33]:
from collections import defaultdict

code_names = defaultdict(list)

for row in vic_postcode_names.iterrows():
    #print(row[1][0], row[1][1])
    code_names[row[1][1]].append(row[1][0])

#code_names['POA3000']

In [34]:
# The code below shows how messy suburbs names are matched to post codes.

show_how_many = 5

for code, name in code_names.items():
    if len(code_names[code]) >= 2:
        print(code, name)
    if show_how_many <=0:
        break
    show_how_many -= 1
        

POA3737 ['Abbeyard', 'Barwidgee', 'Buffalo River', 'Dandongadale', 'Gapsted', 'Havilah', 'Merriang', 'Merriang South', 'Mudgegonga', 'Myrtleford', 'Nug Nug', 'Rosewhite']
POA3040 ['Aberfeldie', 'Essendon', 'Essendon West']
POA3825 ['Aberfeldy', 'Amor', 'Caringal', 'Coalville', 'Erica', 'Fumina', 'Fumina South', 'Hernes Oak', 'Hill End', 'Jacob Creek', 'Moe', 'Moe South', 'Moondarra', 'Newborough', 'Rawson', 'Tanjil South', 'Thalloo', 'Walhalla', 'Westbury', 'Willow Grove', 'Yallourn', 'Yallourn North']
POA3714 ['Acheron', 'Alexandra', 'Cathkin', 'Devils River', 'Fawcett', 'Koriella', 'Maintongoon', 'Taggerty', 'Whanregarwen']
POA3984 ['Adams Estate', 'Caldermeade', 'Corinella', 'Coronet Bay', 'Grantville', 'Jam Jerrup', 'Lang Lang', 'Lang Lang East', 'Monomeith', 'Pioneer Bay', 'Queensferry', 'Tenby Point', 'The Gurdies']


In [35]:
vic_postal_income = vic_postal_income.reset_index()
vic_postal_income['suburb'] = vic_postal_income['postcode'].apply(lambda row: code_names[row])

## Top average weekly income suburbs

In [36]:
vic_postal_income.nlargest(10, 'avg_income')

postcode  number  total_income   avg_income  \
556  POA3785      28       29764.0  1063.000000   
471  POA3662     771      747619.0   969.674449   
600  POA3852      72       69268.5   962.062500   
61   POA3067    7525     6738170.5   895.437940   
163  POA3184   13411    12000147.0   894.798822   
103  POA3121   27444    24471396.5   891.684758   
1    POA3002    4631     4082104.5   881.473656   
11   POA3015   13837    12134495.5   876.959999   
497  POA3704      17       14891.5   875.970588   
10   POA3013   12168    10538549.5   866.087237   

                                     suburb  
556                               [Tremont]  
471                           [Puckapunyal]  
600                             [East Sale]  
61                             [Abbotsford]  
163                                [Elwood]  
103           [Burnley, Cremorne, Richmond]  
1                          [East Melbourne]  
11   [Newport, South Kingsville, Spotswood]  
497                               [Koetong]  
10                             [Yarraville]

## Top total population weekly income suburbs

In [37]:
vic_postal_income.nlargest(10, 'total_income')

postcode  number  total_income  avg_income  \
25   POA3030   75834    54920404.0  724.218741   
685  POA3977   70605    48324142.0  684.429460   
24   POA3029   70398    47010874.5  667.787075   
58   POA3064   55877    34154114.5  611.237441   
132  POA3150   50230    33846014.0  673.820705   
285  POA3350   48124    32833054.0  682.259455   
191  POA3216   46183    32115504.0  695.396661   
18   POA3023   48665    31430910.0  645.862735   
178  POA3199   44887    30262235.5  674.187081   
572  POA3805   44104    29381543.5  666.187727   

                                                suburb  
25   [Derrimut, Point Cook, Quandong, Werribee, Wer...  
685  [Botanic Ridge, Cannons Creek, Cranbourne, Cra...  
24              [Hoppers Crossing, Tarneit, Truganina]  
58   [Craigieburn, Donnybrook, Kalkallo, Mickleham,...  
132                     [Glen Waverley, Wheelers Hill]  
285  [Alfredton, Bakery Hill, Ballarat Central, Bal...  
191  [Belmont, Grovedale, Highton, Marshall, Wandan...  
18   [Burnside, Burnside Heights, Cairnlea, Carolin...  
178                       [Frankston, Frankston South]  
572                 [Narre Warren, Narre Warren South]

## Bottom average weekly income suburbs

In [38]:
vic_postal_income.nsmallest(10, 'avg_income')

postcode  number  total_income  avg_income  \
77   POA3086     881      145962.5  165.678207   
568  POA3800    2180      398334.5  182.722248   
44   POA3050      76       14292.0  188.052632   
7    POA3010    1581      316896.5  200.440544   
386  POA3520     135       37603.0  278.540741   
378  POA3506      15        4422.0  294.800000   
39   POA3045      86       25743.0  299.337209   
21   POA3026      40       12412.5  310.312500   
623  POA3889     124       39204.0  316.161290   
364  POA3482     111       38861.0  350.099099   

                                                suburb  
77                                                  []  
568                                                 []  
44                                                  []  
7                                                   []  
386                          [Kinypanial, Korong Vale]  
378                                         [Cowangie]  
39                                 [Melbourne Airport]  
21                                    [Laverton North]  
623  [Bellbird Creek, Bemm River, Cabbage Tree Cree...  
364  [Massey, Morton Plains, Warmur, Watchem, Watch...

## Bottom total population weekly income suburbs

In [39]:
vic_postal_income.nsmallest(10, 'total_income')

postcode  number  total_income  avg_income  \
56   POA3062       0           0.0         NaN   
378  POA3506      15        4422.0  294.800000   
322  POA3415      20        7168.0  358.400000   
627  POA3893      13        7543.5  580.269231   
466  POA3647      23        9070.0  394.347826   
500  POA3708      21       11091.0  528.142857   
21   POA3026      40       12412.5  310.312500   
44   POA3050      76       14292.0  188.052632   
369  POA3489      33       14464.5  438.318182   
497  POA3704      17       14891.5  875.970588   

                               suburb  
56                         [Somerton]  
378                        [Cowangie]  
322                           [Miram]  
627  [Double Bridges, Tambo Crossing]  
466                  [Dookie College]  
500                       [Tintaldra]  
21                   [Laverton North]  
44                                 []  
369                           [Tempy]  
497                         [Koetong]

In [40]:
postcode_lowest_total_pop_income = vic_postal_income.nsmallest(10, 'total_income')['postcode']
postcode_lowest_avg_pop_income = vic_postal_income.nsmallest(10, 'avg_income')['postcode']
postcode_top_total_pop_income = vic_postal_income.nlargest(10, 'total_income')['postcode']
postcode_top_avg_pop_income = vic_postal_income.nlargest(10, 'avg_income')['postcode']

In [41]:
top_fa_list = []
for code in postcode_top_avg_pop_income:
    #print(vic_f[vic['postcode'] == code])
    top_fa_list.append(vic_f[vic['postcode'] == code])
top_fa = pd.concat(top_fa_list)
top_fa.head()

/Users/apple/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  after removing the cwd from sys.path.


postcode gender  age_low  age_high  income_low  income_high  number  \
73846  POA3785      F       15        19         1.0        149.0       0   
74544  POA3785      F       20        24         1.0        149.0       0   
75242  POA3785      F       25        34         1.0        149.0       0   
75940  POA3785      F       35        44         1.0        149.0       0   
76638  POA3785      F       45        54         1.0        149.0       0   

       income  total_income  
73846    75.0           0.0  
74544    75.0           0.0  
75242    75.0           0.0  
75940    75.0           0.0  
76638    75.0           0.0

In [42]:
top_ma_list = []
for code in postcode_top_avg_pop_income:
    #print(vic_f[vic['postcode'] == code])
    top_ma_list.append(vic_m[vic['postcode'] == code])
top_ma = pd.concat(top_ma_list)
top_ma.head()


/Users/apple/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  after removing the cwd from sys.path.


postcode gender  age_low  age_high  income_low  income_high  number  \
556   POA3785      M       15        19         1.0        149.0       0   
1254  POA3785      M       20        24         1.0        149.0       0   
1952  POA3785      M       25        34         1.0        149.0       0   
2650  POA3785      M       35        44         1.0        149.0       0   
3348  POA3785      M       45        54         1.0        149.0       0   

      income  total_income  
556     75.0           0.0  
1254    75.0           0.0  
1952    75.0           0.0  
2650    75.0           0.0  
3348    75.0           0.0

In [43]:
# I can't tell much from the female vs male population in the top suburbs.
top_fa['number'].sum(), top_ma['number'].sum()

(41063, 38653)

In [44]:
# I couldn't find an applicable case for crosstab now
# pd.crosstab(age2['GROUPED_AGE'], age2['AGE'])

# Sum up

Replacing missing values should make the data analysis more accurate and show a real picture.

Here are the missing data:
* negative income -1  in the “income_low” column
* unknown income -99  in the “income_low” column
* unknown income -99  in the “income_high” column
* unknown income 9999  in the “income_high” column
All are changed to nump.nan


When using qcut to explore the distribution of population, it’s discovered that the majority of population is fewer or equal to 3, accounting for nearly 50%. The reason is that Victoria has a relatively smaller population so the population per post code per income bracket is small. That can have an impact on analysis based on post code.

I also created secondary variables and discover some meaningful facts:

* “income” is created by averaging the lower and higher ends of the income brackets.
* “total_income” is created by timing “income” and “number” together. We made an assumption that the average income is the average of the lower and higher ends.
* Using the secondary variables, we can quickly tell the average weekly income ratio is, Female : Male = 591.53 : 773.68

To further explore the influence of post code (location), I extract top and bottom 10 income post codes. Since it’s difficult to use the post code alone, I try to map the code to suburb names. In the process, I realise the mapping is not perfect but I try it anyway because it is far more easier to interpreter the data by name.

The analysis is not yet complete and I hope some data visualization techniques can help me discover more interesting areas I can use for further investigation next week.
